src/deepsparse/timing/timing_builder.py

```diff
@@ -0,0 +1,133 @@
+
+from deepsparse.timing.timing_schema import InferenceTimingSchema
+
+
+__all__ = ["TimingBuilder"]
+
+
+class TimingBuilder:
+    """
+    This object aggregates the durations
+    (time deltas in seconds) of various components of the inference
+    pipeline.
+    Once all the desired time deltas are aggregated, the
+    object may build the data structure that summarizes the
+    collected information.
+
+    Example flow:
+
+    ```
+    builder = TimingBuilder()
+    builder.initialize()
+
+    builder.start("total_inference")
+
+    builder.start("pre_process")
+    do_something()
+    builder.stop("pre_process")
+
+    builder.start("engine_forward")
+    do_something()
+    builder.stop("engine_forward")
+
+    builder.start("post_process")
+    do_something()
+    builder.stop("post_process")
+
+    builder.stop("total_inference")
+    ...
+    summary = builder.build()
+    ```
+    The object may time the duration of an arbitrary number
+    of events (phases), but the names of phases need to
+    be consistent with the `schema` argument passed to
+    the build() method
+    """
+
+    def __init__(self):
+        self.initialized = False
+        self._start_stop_times = {}
+
+    def start(self, phase_name: str):
+        """
+        Collect the starting time of the phase
+
+        :param phase_name: The name of an event (phase), which duration
+            we are measuring
+        """
+        if not self.initialized:
+            raise ValueError(
+                "Attempting to collect time information, "
+                "but the TimingBuilder instance not initialized. "
+                "Call initialize() method first"
+            )
+        if phase_name in self._start_stop_times:
+            raise ValueError(
+                f"Attempting to overwrite the start time of the phase: {phase_name}"
+            )
+        self._start_stop_times[phase_name] = {"start": time.time()}
+
+    def stop(self, phase_name: str):
+        """
+        Collect the finish time of the phase
+
+        :param phase_name: The name of an event (phase), which duration
+            we are measuring
+        """
+        if phase_name not in self._start_stop_times:
+            raise ValueError(
+                f"Attempting to grab the stop time of the phase: {phase_name},"
+                f"but its start time is missing"
+            )
+        if "stop" in self._start_stop_times[phase_name]:
+            raise ValueError(
+                f"Attempting to overwrite the stop time of the phase: {phase_name}"
+            )
+        self._start_stop_times[phase_name]["stop"] = time.time()
+
+    def initialize(self):
+        """
+        Prime the builder, so it is ready to collect time measurements
+        """
+        if self.initialized:
+            raise ValueError("The TimingBuilder instance has been already initialized")
+        self.initialized = True
+
+    def build(self, schema: BaseModel = InferenceTimingSchema) -> BaseModel:
```

Comment:
What's the reasoning for using pydantic here? This is all an internal api, so assuming its tested, it should never create fields that shouldn't be there. 

An alternative to this approach is to have a `def query(self, phase) -> time` method. Then you wouldn't have to return anything and wouldn't need this extra step.

Reply:
Not entirely sure what you mean here. My goal was to use TimerBuilder to be able to:
-  create time deltas for phases with different names
-  once the deltas area is created, put them into a schema to be returned for further consumption

I feel that using Pydantic schema here prevents us from e.g. having a typo in one of our `phase_name`s.
E.g. `pre_processing` could be erronously written as `pre_processing_time`.  Pydantic would catch that.

Reply:
Ahh composition gotcha, that makes sense. Right cause these will be summed/averaged together from the different runs? Okay yeah this makes more sense then!